openhands/controller/agent_controller.py

```diff
@@ -133,21 +133,19 @@ async def update_state_after_step(self):
         # update metrics especially for cost. Use deepcopy to avoid it being modified by agent.reset()
         self.state.local_metrics = copy.deepcopy(self.agent.llm.metrics)
 
-    async def report_error(self, message: str, exception: Exception | None = None):
-        """Reports an error to the user and sends the exception to the LLM next step, in the hope it can self-correct.
-
-        This method should be called for a particular type of errors, which have:
-        - a user-friendly message, which will be shown in the chat box. This should not be a raw exception message.
-        - an ErrorObservation that can be sent to the LLM by the user role, with the exception message, so it can self-correct next time.
-        """
-        self.state.last_error = message
-        if exception:
-            self.state.last_error += f': {exception}'
+    async def _react_to_error(
+        self,
+        message: str,
+        exception: Exception | None = None,
+        new_state: AgentState | None = None,
+    ):
+        if new_state is not None:
+            await self.set_agent_state_to(new_state)
         detail = str(exception) if exception is not None else ''
         if exception is not None and isinstance(exception, litellm.AuthenticationError):
-            detail = 'Please check your credentials. Is your API key correct?'
-        self.event_stream.add_event(
-            ErrorObservation(f'{message}:{detail}'), EventSource.USER
+            detail += '\nPlease check your credentials. Is your API key correct?'
+        await self.event_stream.async_add_event(
+            ErrorObservation(f'{message}:{detail}'), EventSource.AGENT
```

Comment:
This method was initially meant only for informing the user in the UI. But it had such a good name, that we felt like it should be used almost like logging errors. 😂

Reply:
😂 

TBH I don't think an AgentController should know anything about a UI

Reply:
Well, it's not that it knows it, strictly speaking, just the ErrorObs created here has a message intended for the chat box / end user's eyes.

And by this time, it doesn't have anymore the info about the exception, except... it did, when it got it as param for some exception info intended for evals.

Reply:
Actually, you're right. How about:
- we don't do at all these UI-intended lines of code here, in the controller, 
- create an ErrorObservation intended for the backend / LLM
- with exception information
- put it in the stream
- let the agent session interpret it and send over whatever info for the UI it makes of it.